### PR TITLE
fix(route): changba user

### LINKED ISSUE #7470: 唱吧用户页面变更，路由报错
[Issuehunt badges](https://issuehunt.io/r/DIYgod/RSSHub/issues/7470)

#### 路由地址（不包含参数）
路由: /changba/:userid
#### 完整路由地址，包含所有必选与可选参数
https://rsshub.app/changba/34108440
#### 相关文档地址
https://docs.rsshub.app/social-media.html#chang-ba
#### 预期是什么
正常抓取用户作品
#### 实际发生了什么
报错 empty route
#### 部署相关信息
官方 Demo 及 Docker 自建（最新版）
#### 额外信息（日志、报错等）

唱吧官方变更了用户页面地址，（可能页面内容也有变化，不确定）
新的用户页面地址示例：
https://changba.com/wap/index.php?s=skp6hhF59n48R-UpqO3izw

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $2.00)

- [yefoenix yefoenix](https://issuehunt.io/u/yefoenix) ($2.00)

### Submitted pull Requests
- [#7538 fix(route): changba user](https://issuehunt.io/r/DIYgod/RSSHub/pull/7538)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/DIYgod/RSSHub/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by xizeyoupan**
## 该 PR 相关 Issue / Involved issue

Close #7470

## 完整路由地址 / Example for the proposed route(s)

```routes
/changba/skp6hhF59n48R-UpqO3izw
```

## 新RSS检查列表 / New RSS Script Checklist

- [ ] 这个PR中包含了新的路由吗? Does this PR add new route?
  - 如果有, 请完成检查列表. If yes, please finish the check list
  - **如果你的PR符合下方某个事项, 也请注明. If any of the checklist item meets your PR, please fill it out.**
  - [x] <- 这样打勾
- [x] 是否提供了文档? Documentation provided?
  - [ ] 是否提供了英文文档? EN Documentation provided?
- [ ] 是否支持全文获取? Is this RSS Script support fulltext?
  - [ ] 如果全文获取中需要访问文章链接, 是否使用了缓存? If fulltext requires to fetch detail pages, is cache used in the process?
  - [缓存说明](https://docs.rsshub.app/joinus/quick-start.html#ti-jiao-xin-de-rsshub-gui-ze-bian-xie-jiao-ben-shi-yong-huan-cun) | [How to use cache](https://docs.rsshub.app/joinus/quick-start.html#ti-jiao-xin-de-rsshub-gui-ze-bian-xie-jiao-ben-shi-yong-huan-cun)
- [ ] 目标是否有明显的反爬/频率限制? Is there any sign of anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? (延长缓存时间, 写文档说明, etc.) If yes, do your code reflect this sign? (e.g. write documentations, use long cache time)
- [ ] 目标是否有提供日期？ Is there a date in the source?
  - [ ] 如果有，包是否正确解析？ If there is, can this script provide this info?
  - [ ] 如果有提供解析能力，时区是否正确调整？ Is the timezone correctly provided?
  - 如果有提供日期，但是没有提供解析，请说明原因 If there is a date but this script does not parse, please provide your reason.
- [ ] 是否引入的新的包? Any new package introduced?
  - 如果有, 请说明原因. If yes, please state your reason
- [ ] 是否使用了`Puppeteer`? Make use of `Puppeteer`?
  - 如果有, 请说明原因. If yes, please state your reason
  

## 说明 / Note

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#7470: 唱吧用户页面变更，路由报错](https://issuehunt.io/repos/127769231/issues/7470)
---
</details>

### CODE REVIEW

File: lib/routes/changba/user.js [hunk 1]
```diff
@@ -1,60 +1,54 @@
 const got = require('@/utils/got');
 const cheerio = require('cheerio');
+const headers = { 'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1' };
```
Comment (SettingDust) [inline lib/routes/changba/user.js:3]: 使用默认 ua 无法访问？

File: lib/routes/changba/user.js [hunk 2]
```diff
@@ -1,60 +1,54 @@
...
-
     const items = await Promise.all(
         list.map(async (item) => {
             const $ = cheerio.load(item);
             const link = $('a').attr('href');
+            const songid = await ctx.cache.tryGet(
+                link,
+                async () => {
+                    const result = await got.get({
+                        method: 'get',
```
Comment (SettingDust) [inline lib/routes/changba/user.js:27]: 写两个 get？
```suggestion
                    const result = await got({
                        method: 'get',
```

File: lib/routes/changba/user.js [hunk 3]
```diff
@@ -1,60 +1,54 @@
...
+                    const re = /workid: '\d+'/;
+                    let workid = result.data.match(re)[0];
+                    workid = workid.split("'")[1];
+                    return workid;
                 },
-            });
-
-            const result = ProcessFeed(response.data);
-
+                60 * 60 * 24
```
Comment (SettingDust) [inline lib/routes/changba/user.js:37]: 不使用默认过期时间的原因是？

File: lib/routes/changba/user.js [hunk 4]
```diff
@@ -1,60 +1,54 @@
...
                 author: author,
-                itunes_item_image: authorimg,
-                enclosure_url: result.mp3,
+                itunes_item_image: itunes_item_image,
+                enclosure_url: mp3,
                 enclosure_type: 'audio/mpeg',
             };
-            ctx.cache.set(link, JSON.stringify(single));
+
             return Promise.resolve(single);
```
Comment (SettingDust) [inline lib/routes/changba/user.js:52]: 这种写法应该过不了 lint 吧？

```suggestion
            return {
                title: $('.work-title').text(),
                description,
                link,
                author,
                itunes_item_image,
                enclosure_url: mp3,
                enclosure_type: 'audio/mpeg',
            };
```

**Review by SettingDust — CHANGES_REQUESTED** [review]

File: lib/routes/changba/user.js (hunk 1 above)
Comment (xizeyoupan) [reply lib/routes/changba/user.js:3]: 确实

File: lib/routes/changba/user.js (hunk 2 above)
Comment (xizeyoupan) [reply lib/routes/changba/user.js:27]: typo😳

File: lib/routes/changba/user.js (hunk 3 above)
Comment (xizeyoupan) [reply lib/routes/changba/user.js:37]: 获取歌曲id，理论上不会更新

File: lib/routes/changba/user.js (hunk 4 above)
Comment (xizeyoupan) [reply lib/routes/changba/user.js:52]: 远古 PR 了属于是，lint 已经改了好多次了。
image

**Comment by xizeyoupan** [comment]
@DIYgod 您好！~~由于总是先处理最新的，~~ 这个远古PR当年不知道为什么迷失在了时间乱流中，现在我已经迁移到了v2，如果没问题的话，斗胆@您有空merge一下。

File: lib/v2/changba/user.js [hunk 5]
```diff
@@ -0,0 +1,86 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const { art } = require('@/utils/render');
+const path = require('path');
+const headers = { 'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 11_0 like Mac OS X) AppleWebKit/604.1.38 (KHTML, like Gecko) Version/11.0 Mobile/15A372 Safari/604.1' };
```
Comment (TonyRL) [inline lib/v2/changba/user.js:5]: RSSHub's UA can also be changed via envs.
https://github.com/DIYgod/RSSHub/blob/bdb0e125d28ee4048418ded3ccd083feedbb326e/lib/config.js#L40

Consider using that value if envs.UA is set.

**Review by TonyRL — CHANGES_REQUESTED** [review]

File: lib/v2/changba/radar.js [hunk 6]
```diff
@@ -0,0 +1,13 @@
+module.exports = {
+    'changba.com': {
+        _name: '唱吧',
+        '.': [
+            {
+                title: '用户',
+                docs: 'https://docs.rsshub.app/reading.html#chang-ba',
+                source: ['/s/:id'],
+                target: `/changba/:id`,
```
Comment (TonyRL) [inline lib/v2/changba/radar.js:9]:
```suggestion
                docs: 'https://docs.rsshub.app/social-media.html#chang-ba',
                source: ['/s/:userid'],
                target: `/changba/:userid`,
```

File: lib/v2/changba/user.js (hunk 5 above)
Comment (TonyRL) [inline lib/v2/changba/user.js:5]: RSSHub's UA can also be changed via envs.
https://github.com/DIYgod/RSSHub/blob/bdb0e125d28ee4048418ded3ccd083feedbb326e/lib/config.js#L40
Consider using that value if `envs.UA` is set.

File: lib/v2/changba/user.js [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+                    workid = workid.split("'")[1];
+                    return workid;
+                },
+                60 * 60 * 24
+            );
+            if (!songid) {
+                return null;
+            }
+            const mp3 = `http://upscuw.changba.com/${songid}.mp3`;
```
Comment (TonyRL) [inline lib/v2/changba/user.js:55]:
```suggestion
            const mp3 = `https://upscuw.changba.com/${songid}.mp3`;
```
HTTPS is prefered

File: lib/v2/changba/user.js [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                title: $('.work-title').text(),
+                description,
+                link,
+                author,
+                itunes_item_image,
+                enclosure_url: mp3,
+                enclosure_type: 'audio/mpeg',
+            };
+
+            return Promise.resolve(single);
```
Comment (TonyRL) [inline lib/v2/changba/user.js:71]: See https://github.com/DIYgod/RSSHub/pull/7538#discussion_r701082982

**Review by TonyRL — CHANGES_REQUESTED** [review]

**Pull request merged**